micrometer-jetty12/src/main/java/io/micrometer/jetty12/client/JettyClientKeyValues.java

```diff
@@ -0,0 +1,163 @@
+                : STATUS_UNKNOWN;
+    }
+
+    /**
+     * Creates a {@code uri} KeyValue based on the URI of the given {@code result}.
+     * {@code REDIRECTION} for 3xx responses, {@code NOT_FOUND} for 404 responses.
+     * @param request the request
+     * @param result the request result
+     * @param successfulUriPattern successful URI pattern
+     * @return the uri KeyValue derived from the request and its result
+     */
+    public static KeyValue uri(Request request, @Nullable Result result,
+            BiFunction<Request, Result, String> successfulUriPattern) {
+        if (result != null && result.getResponse() != null) {
+            int status = result.getResponse().getStatus();
+            if (HttpStatus.isRedirection(status)) {
+                return URI_REDIRECTION;
+            }
+            if (status == 404) {
+                return URI_NOT_FOUND;
+            }
+        }
+
+        String matchingPattern = successfulUriPattern.apply(request, result);
+        matchingPattern = MULTIPLE_SLASH_PATTERN.matcher(matchingPattern).replaceAll("/");
+        if (matchingPattern.equals("/")) {
+            return URI_ROOT;
+        }
+        matchingPattern = TRAILING_SLASH_PATTERN.matcher(matchingPattern).replaceAll("");
+        return KeyValue.of("uri", matchingPattern);
+    }
+
+    /**
+     * Creates an {@code exception} KeyValue based on the {@link Class#getSimpleName()
+     * simple name} of the class of the given {@code exception}.
+     * @param result the request result
+     * @return the exception KeyValue derived from the exception
+     */
+    public static KeyValue exception(@Nullable Result result) {
+        if (result == null) {
+            return EXCEPTION_UNKNOWN;
+        }
+        Throwable exception = result.getFailure();
+        if (exception == null) {
+            return EXCEPTION_NONE;
+        }
+        if (result.getResponse() != null) {
+            int status = result.getResponse().getStatus();
+            if (status == 404 || HttpStatus.isRedirection(status)) {
+                return EXCEPTION_NONE;
+            }
+        }
+        if (exception.getCause() != null) {
+            exception = exception.getCause();
+        }
+        String simpleName = exception.getClass().getSimpleName();
+        return KeyValue.of("exception",
+                StringUtils.isNotEmpty(simpleName) ? simpleName : exception.getClass().getName());
+    }
+
+    /**
+     * Creates an {@code outcome} KeyValue based on the status of the given
+     * {@code result}.
+     * @param result the request result
+     * @return the outcome KeyValue derived from the status of the response
+     */
+    public static KeyValue outcome(@Nullable Result result) {
+        if (result == null) {
+            return OUTCOME_UNKNOWN;
+        }
+        return Outcome.forStatus(result.getResponse().getStatus()).asKeyValue();
+    }
```

Comment:
With HTTP/1 the existence of the status code is just what the response commit buffer had, not that the request/response is complete or successful (depending on the mode used in HTTP/1 there could be a socket exception, or even just a normal socket closure with incomplete data see "close-delimited message" on the HTTP/1 specs). 

With HTTP/2 and HTTP/3 the status code is just what was seen on the first buffer.
Those responses can also be canceled via the HTTP/2 and HTTP/3 protocols and be incomplete, these would be seen as protocol specific exceptions.

Reply:
Those are interesting scenarios and I don't think we're properly handling them right now. It sounds like something we could write a test for and then figure out how to (or if we can) improve all the different HTTP instrumentations (/cc @bclozel). For HTTP client instrumentation we have a small suite of tests that we run all of them against; see [HttpClientTimingInstrumentationVerificationTests](https://github.com/micrometer-metrics/micrometer/blob/63cfcbccd57dabb9cfffe77b613626ded47803d2/micrometer-test/src/main/java/io/micrometer/core/instrument/HttpClientTimingInstrumentationVerificationTests.java) which is extended in this pull request to run the instrumentation against it in Jetty12ClientTimingInstrumentationVerificationTests.

Are there existing tests in Jetty for this scenario that we could get inspiration from to add to our test suite?

We haven't done a good job of testing with HTTP/2 or HTTP/3 so far, so that's another point of improvement.